tee-worker/service/src/main_impl.rs

```diff
@@ -495,120 +495,112 @@ fn start_worker<E, T, D, InitializationHandler, WorkerModeProvider>(
+	//
+	// The logic here is as follows:
+	// in the case it's Some, it means there is a primary worker registered for the shard, we are a non-primary worker if it is first run
+	// so we need to request provisioning. Finally in any case we register worker.
+	//
+	// in case it's None, it means there is no primary worker registered for the shard, then we are the primary worker
+	//
+	// based on the check of `enclave.get_shard_creation_info` to tell if this worker has run before - this is similar to upstream.
+	// There're a few cases:
+	// 1. `--clean-reset` is set, then the shard should have been initalized earlier already and it's empty state anyway
+	// 2. `--clean-reset` is not set:
+	//    2a. `get_shard_creation_info` is empty and we are primary worker => it's never run before => init everything
+	//    2b. `get_shard_creation_info` is empty and we are non-primary worker => it's never run before => request to sync state
+	//    2c. `get_shard_creation_info` is non-empty it's run before => do nothing
+	let primary_enclave_id = litentry_rpc_api
+		.primary_enclave_identifier_for_shard(WorkerType::Identity, shard, None)
+		.expect("primary_enclave_identifier_for_shard failed");
+
+	let mut register_enclave_xt_header: Header;
+
+	let (we_are_primary_validateer, re_init_parentchain_needed) = match primary_enclave_id {
+		Some(primary_account_id) => {
+			info!("Primary enclave for shard is: {:?}", primary_account_id);
+			let is_first_run = enclave
+				.get_shard_creation_info(shard)
+				.unwrap()
+				.for_parentchain(ParentchainId::Litentry)
+				.is_none();
+			let we_are_primary_validateer = primary_account_id == tee_accountid;
+
+			if is_first_run && !we_are_primary_validateer {
+				// we are not primary worker
+				info!("my state doesn't know the creation header of the shard. will request provisioning");
+				// obtain provisioning from last active worker as this hasn't been done before
+				sync_state::sync_state::<_, _, WorkerModeProvider>(
+					&litentry_rpc_api,
+					&shard,
+					enclave.as_ref(),
+					skip_ra,
+				);
+			}
+			// we register the enclave to have the keys up to date after provisioning
+			let register_enclave_block_hash = register_enclave(
+				enclave.clone(),
+				&litentry_rpc_api,
+				&tee_accountid,
+				&trusted_url,
+				skip_ra,
+				is_development_mode,
+			)
+			.expect("enclave RA registration must be successful to continue");
 
-	let api_register_enclave_xt_header =
-		litentry_rpc_api.get_header(Some(register_enclave_block_hash)).unwrap().unwrap();
+			register_enclave_xt_header =
+				get_registered_enclave_xt_header(&litentry_rpc_api, register_enclave_block_hash);
 
-	// TODO: #1451: Fix api-client type hacks
-	let register_enclave_xt_header =
-		Header::decode(&mut api_register_enclave_xt_header.encode().as_slice())
-			.expect("Can decode previously encoded header; qed");
+			(we_are_primary_validateer, is_first_run)
```

Comment:
Overall looks good to me 

One small distinction to the original impl is:
for non-primary worker and not the first run, we'll return `(false, false)` here.

In the old impl we'll never have such value, so `re_init_parentchain_needed` will always be true for non-primary workers.

Shall we change it to the same logic as before?

Reply:
Good catch, I'll change `re_init_parentchain_needed` to follow the previous logic